File: ServiGo/backend/apps/usuarios/repositories/usuario_repository.py
from django.contrib.auth import authenticate

from ..data.roles import RolUsuario
from ..models import Usuario


class UsuarioRepository:
    """Acceso a datos para Usuario. No contiene reglas de negocio,
    solo lectura/escritura contra la base de datos."""

    @staticmethod
    def listar_todos():
        return Usuario.objects.all()

    @staticmethod
    def buscar_por_id(usuario_id):
        return Usuario.objects.get(pk=usuario_id)

    @staticmethod
    def autenticar(username, password):
        return authenticate(username=username, password=password)

    @staticmethod
    def clientes():
        return Usuario.objects.filter(rol=RolUsuario.CLIENTE)

    @staticmethod
    def clientes_de_profesional(profesional_id):
        return Usuario.objects.filter(
            rol=RolUsuario.CLIENTE,
            reservas__profesional_id=profesional_id
        ).distinct()

File: backend/src/repositories/usuario.repository.ts
import { Usuario, IUsuario } from "../models/Usuario";
import { ROLES } from "../data/roles";

export class UsuarioRepository {
  static listarTodos() {
    return Usuario.find().select("-password").lean();
  }
  static buscarPorId(id: string) {
    return Usuario.findById(id).select("-password");
  }
  static buscarPorUsername(username: string, incluirPassword = false) {
    const query = Usuario.findOne({ username });
    return incluirPassword ? query.select("+password") : query;
  }
  static clientes() {
    return Usuario.find({ rol: ROLES.CLI }).select("-password");
  }
  static clientesDeProfesional(profesionalId: string) {
    return Usuario.find({
      rol: ROLES.CLI,
      _id: { $in: [] }
    }).select("-password");
  }
  static crear(data: Partial<IUsuario>) {
    return Usuario.create(data);
  }
}
